presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.type;
+
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+import java.util.Objects;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static java.lang.String.format;
+import static java.util.Collections.singletonList;
+
+public final class CharType
+        extends AbstractVariableWidthType
+{
+    public static final int MAX_LENGTH = 65_536;
+
+    public static CharType createCharType(long length)
+    {
+        return new CharType(length);
+    }
+
+    private final int length;
+
+    private CharType(long length)
+    {
+        super(
+                new TypeSignature(
+                        StandardTypes.CHAR,
+                        singletonList(TypeSignatureParameter.of(length))),
+                Slice.class);
+
+        if (length < 1 || length > MAX_LENGTH) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("CHAR length scale must be in range [1, %s]", MAX_LENGTH));
+        }
+        this.length = (int) length;
+    }
+
+    public int getLength()
+    {
+        return length;
+    }
+
+    @Override
+    public boolean isComparable()
+    {
+        return true;
+    }
+
+    @Override
+    public boolean isOrderable()
+    {
+        return true;
+    }
+
+    @Override
+    public Object getObjectValue(ConnectorSession session, Block block, int position)
+    {
+        if (block.isNull(position)) {
+            return null;
+        }
+
+        return block.getSlice(position, 0, block.getLength(position)).toStringUtf8();
+    }
+
+    @Override
+    public boolean equalTo(Block leftBlock, int leftPosition, Block rightBlock, int rightPosition)
+    {
+        int leftLength = leftBlock.getLength(leftPosition);
+        int rightLength = rightBlock.getLength(rightPosition);
+        if (leftLength != rightLength) {
+            return false;
+        }
+        return leftBlock.equals(leftPosition, 0, rightBlock, rightPosition, 0, leftLength);
```

Comment:
> 2) All code in Presto, after analysis, must assume that type only coercions may be implicitly present.

Do you mean that type only coercion makes two types interchangeable? While their representation is the same, it doesn't necessarily mean that their semantics is the same too.

> I see a few ways forward here:

We will try some combination of 1) and 2) since it will let us merge the code faster. Maybe the best would be to add `experimental.enable-char-operators` flag so that the `CHAR` data can be just pipelined through Presto.
We are also happy to fix type system issues.

> but one thing we should test first is whether those functions are actually necessary

In case of primitive types those functions are probably not required. However, there will be performance impact in `Slice` based types, since `Slice` instances would have to be instantiated for each `EQUAL` call. That is why I think `Block` based function that are registered in `TypeManager` would be the best approach.